### PR TITLE
RISC-V: New CPU (FE310) and Board (HiFive1) 

### PR DESCRIPTION
**Pull request by kenrabold**
Updated PR for supporting new RISC-V CPU and board from SiFive.

This PR replaces PR [#8275](https://github.com/RIOT-OS/RIOT/pull/8275) and addresses all comments and recommendations listed there.

This port leverages vendor header and source files from SiFive, which are in /vendor/ and /include/vendor folders.  RIOT CI scripts have been updated to ignore formatting in those vendor supplied files.

### CODE REVIEW

**Comment by kaspar030** [comment]
Seems github has some trouble accepting pushes.

**Comment by kaspar030** [comment]
@kenrabold Thanks a lot for taking over!

Looking at Murdock's results,

- for some tests and examples, the hifive needs to be added to ```BOARD_INSUFFICIENT_MEMORY```

- some syscalls seem to be missing (getpid, kill, abort). Maybe risc-v can somehow re-use the default syscalls? The code looks the same, apart from the reent-parameters

**Comment by kaspar030** [comment]
I'll create a PR for the missing stdio include in uhcp...

**Comment by kaspar030** [comment]
Fixed in #8967.

@all any suggestion on how to bypass the Doxygen errors in context_frame.h?
Would ```#ifndef DOXYGEN``` work?

**Comment by dylad** [comment]
@kaspar030 You probably wanted to ask the question to @RIOT-OS/maintainers ?

**Comment by miri64** [comment]
> [@RIOT-OS/maintainers] any suggestion on how to bypass the Doxygen errors in context_frame.h?
Would #ifndef DOXYGEN work?

Move to `cpu/include/vendor`?

**Comment by gebart** [comment]
Is it a vendor file? Do what @miri64 said.
Is it not a vendor file? Correct the Doxygen errors/warnings in the file.

**Comment by kaspar030** [comment]
> Is it not a vendor file? Correct the Doxygen errors/warnings in the file.

I was hoping someone else would feel that documenting the registers is a little pointless.

@kenrabold seems like we'll have to...

File: cpu/fe310/include/context_frame.h [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    uint32_t a3;
+    uint32_t a4;
+    uint32_t a5;
+    uint32_t a6;
+    uint32_t a7;
+};
+#endif /* __ASSEMBLER__ */
+
+
+/* These values are checked for correctness in context_frame.c */
```
Comment (gebart) [inline cpu/fe310/include/context_frame.h]: Add some general docs for what these macros are. You could do a `/* @name something @{ `

File: cpu/fe310/include/context_frame.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+#define a2_OFFSET     104
+#define a3_OFFSET     108
+#define a4_OFFSET     112
+#define a5_OFFSET     116
+#define a6_OFFSET     120
+#define a7_OFFSET     124
+
+#define CONTEXT_FRAME_SIZE (a7_OFFSET + 4)
+
+#define SP_OFFSET_IN_THREAD 0
```
Comment (gebart) [inline cpu/fe310/include/context_frame.h]: Document this macro

File: cpu/fe310/include/context_frame.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+#define a0_OFFSET     96
+#define a1_OFFSET     100
+#define a2_OFFSET     104
+#define a3_OFFSET     108
+#define a4_OFFSET     112
+#define a5_OFFSET     116
+#define a6_OFFSET     120
+#define a7_OFFSET     124
+
+#define CONTEXT_FRAME_SIZE (a7_OFFSET + 4)
```
Comment (gebart) [inline cpu/fe310/include/context_frame.h]: self explanatory macro name, but could just as well have a single line doxygen comment above it.

**Comment by miri64** [comment]
Me neither, sorry :-/. Another alternative (less intrusive into Doxygen's config) is also to put a group around it:

```C
/**
 * @name Register definitions
 * @{
 */
/* ... */
/** @} */

File: cpu/fe310/include/context_frame.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * This also defines context_switch_frame offsets for assembly language.  The
+ * structure is sized to maintain 16 byte stack alignment per the ABI.
+ * https://github.com/riscv/riscv-elf-psabi-doc
+ *
+ */
+#if !defined(__ASSEMBLER__)
+
+/* N.B.: update the definitions below if this changes */
+struct context_switch_frame {
```
Comment (gebart) [inline cpu/fe310/include/context_frame.h]: Add documentation for the struct. You don't need to document each member variable.

**Comment by kenrabold** [comment]
It took a couple tries, but I finally got all the Doxygen warnings fixed.  I hope all is good now.

File: cpu/fe310/context_frame.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
+
```
Comment (kaspar030) [inline cpu/fe310/context_frame.c]: This file and some others have an empty line at the top, which kinda jumps into the eye. Please fix.

File: cpu/fe310/nano/nanostubs.c [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+{
+    (void) tp;
+    (void) tzp;
+    errno = EINVAL;
+    return -1;
+}
+
+int _getpid(void)
+{
+    return 1;
```
Comment (kaspar030) [inline cpu/fe310/nano/nanostubs.c]: this should return ```sched_active_pid```

File: cpu/fe310/include/context_frame.h [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+ * @brief   context_switch_frame stores the registers and PC for a context switch.
+ *
+ * This also defines context_switch_frame offsets for assembly language.  The
+ * structure is sized to maintain 16 byte stack alignment per the ABI.
+ * https://github.com/riscv/riscv-elf-psabi-doc
+ *
+ */
+struct context_switch_frame {
+    uint32_t filler0;               /**< filler to maintain 16 byte alignment */
+    uint32_t filler1;               /**< filler to maintain 16 byte alignment */
```
Comment (gebart) [inline cpu/fe310/include/context_frame.h]: I think it is more commonplace to write the padding as 
```c
uint32_t padding[2];
```
So you don't need to add more lines to adjust the padding.

File: cpu/fe310/include/context_frame.h [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if !defined(__ASSEMBLER__)
+
+/**
+ * @brief   context_switch_frame stores the registers and PC for a context switch.
```
Comment (gebart) [inline cpu/fe310/include/context_frame.h]: you don't have to repeat the struct name in the brief docs, when rendered the name will be on the line above this line so it's just repetition.

File: cpu/fe310/include/context_frame.h [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+#define a7_OFFSET     124
+/** @} */
+
+/**
+ * @brief Size of context switch frame
+ */
+#define CONTEXT_FRAME_SIZE (a7_OFFSET + 4)
+
+/**
+ * @brief Offset of stack pointer in context frame
```
Comment (gebart) [inline cpu/fe310/include/context_frame.h]: is this documentation really correct? is "context frame" referring to `struct context_switch_frame` or is there some kind of mixup? At offset 0 in the struct there is only padding..

Comment (kenrabold) [reply cpu/fe310/include/context_frame.h]: I'll correct that comment.  It is the offset of the sp member of struct _thread

File: cpu/fe310/context_frame.c (hunk 5 above)
Comment (kenrabold) [reply cpu/fe310/context_frame.c]: Fixed that in context_frame.h and context_frame.c

File: cpu/fe310/include/context_frame.h (hunk 7 above)
Comment (kenrabold) [reply cpu/fe310/include/context_frame.h]: Fixed

File: cpu/fe310/include/context_frame.h (hunk 8 above)
Comment (kenrabold) [reply cpu/fe310/include/context_frame.h]: Fixed

File: cpu/fe310/nano/nanostubs.c (hunk 6 above)
Comment (kenrabold) [reply cpu/fe310/nano/nanostubs.c]: I did this:

**Comment by gebart** [comment]
I don't see why there is a need for a CPU specific newlib syscall implementation for this platform. Why are the default syscall implementations not enough? 
What I can see, the only difference is a setvbuf at init to disable buffering, something that is in an open PR for the default syscalls. The custom syscalls also prevent using VFS with the libc file IO functions, because there are no provisions for VFS in this CPU specific implementation unlike there is in the default implementation.

**Comment by kaspar030** [comment]
> Why are the default syscall implementations not enough?

My thoughts exactly, until I realized that the risc-v system calls don't use the reent parameter. Just using our default implementation leads to undefined references. I think the RISC-V newlib is compiled without ```-DREENTRANT_SYSCALLS_PROVIDED```...

IMO this shouldn't block the PR.

**Comment by kenrabold** [comment]
Any other changes that are needed?

**Comment by kaspar030** [comment]
Sorry for the delay. 

I'm getting Wpedantic warnings from the vendor headers. Disabling "-Wpedantic" for the cpu folder works [patch](https://gist.github.com/ef573d5b698610c61aff01d7a9a0452e).

Other than that, please squash!

**Comment by kaspar030** [comment]
(rdcli_simple overflows ram, the hifive needs to be added to BOARD_INSUFFICIENT_MEMORY)

**Comment by kenrabold** [comment]
Do you want me to squash all of the commits into just one, or should I keep some of the main commits (like the first 4) and squash the others?  I'm just not sure what you want to have left.

**Comment by kaspar030** [comment]
> Do you want me to squash all of the commits into just one, or should I keep some of the main commits (like the first 4) and squash the others? I'm just not sure what you want to have left.

Four commits would be good: (please use "make:" instead of "build:")

- the vendor header commit 
- one adding the cpu
- one adding the board
- one fixing the tests / disabling RISC-V where necessary

Thanks!

**Comment by kaspar030** [comment]
Alright, let's go. Many thanks @kenrabold for shaping this up!

**Pull request merged**